### PR TITLE
Migrate scripts to Python3 (part 3)

### PR DESCRIPTION
**Pull request by iarspider**

### CODE REVIEW

File: es_cmsdoxygen_apache.py [hunk 1]
```diff
@@ -1,5 +1,4 @@
-#!/usr/bin/env python
```
Comment (iarspider) [inline es_cmsdoxygen_apache.py:1]: Notice: can't find usage of this script in cms jenkins

File: es_cmsrep_apache.py [hunk 2]
```diff
@@ -1,5 +1,4 @@
-#!/usr/bin/env python
```
Comment (iarspider) [inline es_cmsrep_apache.py:1]: Notice: can't find usage of this script in cms jenkins

File: es_cmssdt_apache.py [hunk 3]
```diff
@@ -1,5 +1,4 @@
-#!/usr/bin/env python
```
Comment (iarspider) [inline es_cmssdt_apache.py:1]: Notice: can't find usage of this script in cms jenkins

File: es_doxygen_apache.py [hunk 4]
```diff
@@ -1,5 +1,4 @@
-#!/usr/bin/env python
```
Comment (iarspider) [inline es_doxygen_apache.py:1]: Notice: can't find usage of this script in cms jenkins

File: es_hypernews.py [hunk 5]
```diff
@@ -1,5 +1,4 @@
-#!/usr/bin/env python
```
Comment (iarspider) [inline es_hypernews.py:1]: Notice: can't find usage of this script in cms jenkins

File: es_hypernews_log.py [hunk 6]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
```
Comment (iarspider) [inline es_hypernews_log.py:1]: Notice: can't find usage of this script in cms jenkins

File: es_relval_log.py [hunk 7]
```diff
@@ -1,5 +1,4 @@
-#!/usr/bin/env python
```
Comment (iarspider) [inline es_relval_log.py:1]: Notice: can't find usage of this script in cms jenkins

File: forward-pull-requests [hunk 8]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
```
Comment (iarspider) [inline forward-pull-requests:1]: Looks like a deprecated script, [last used](https://github.com/search?q=repo%3Acms-sw%2Fcmssw+%22Please+wait+for+a+new+IB%22&type=pullrequests&s=created&o=desc) in 2017

File: github_utils.py [hunk 9]
```diff
@@ -93,6 +93,8 @@ def get_ported_PRs(repo, src_branch, des_branch):
     prRe = re.compile('Automatically ported from ' + src_branch + ' #(\d+)\s+.*', re.MULTILINE)
     for pr in repo.get_pulls(base=des_branch):
         body = pr.body.encode("ascii", "ignore")
+        if sys.version_info[0] == 3:
```
Comment (iarspider) [inline github_utils.py:96]: This function, and `def port_pr`, are only used by `port-pull-request` script, which is not referenced by any job.

File: logwatch.py [hunk 10]
```diff
@@ -38,7 +38,7 @@ def process(self, logs, callback, **kwrds):
           continue
         else: break
       run_cmd ("rsync -a %s %s" % (log, service_log))
-      cur_hash = sha256(run_cmd("head -1 %s" % service_log)).hexdigest()
```
Comment (iarspider) [inline logwatch.py:41]: Common library for es-*-apache scripts (which are unused?)

File: jenkins_monitor_queue.py [hunk 11]
```diff
@@ -30,7 +32,7 @@
 content_hash = get_payload_wscroll('cmssdt-jenkins-queue-*', query_pending_builds)
 if content_hash:
   if (not 'hits' in content_hash) or (not 'hits' in content_hash['hits']):
-    print("ERROR: ", content)
+    print("ERROR: ", content_hash)
```
Comment (iarspider) [inline jenkins_monitor_queue.py:35]: No such variable `content`

File: jenkins_monitor_queue.py [hunk 12]
```diff
@@ -67,7 +69,7 @@
 es_offline_nodes = []
 if content_hash:
   if (not 'hits' in content_hash) or (not 'hits' in content_hash['hits']):
-    print("ERROR: ", content)
+    print("ERROR: ", content_hash)
```
Comment (iarspider) [inline jenkins_monitor_queue.py:72]: No such variable `content`

File: parse_jenkins_builds.py [hunk 13]
```diff
@@ -53,13 +53,13 @@ def process_queue_reason(labels):
     if "already in progress" in labels:
         reason = "concurrent builds not allowed"
     elif "Waiting for next available executor on" in labels:
-        node = labels.split(" on ")[1].encode('ascii', errors='ignore')
-        reason = str(node) + "-busy"
+        node = labels.split(" on ")[1].encode('ascii', errors='ignore').decode("ascii", "ignore")
```
Comment (iarspider) [inline parse_jenkins_builds.py:56]: @aandvalenzuela please review

File: process-error-reports [hunk 14]
```diff
@@ -1,12 +1,11 @@
-#!/usr/bin/env python
+#!/usr/bin/env python3
```
Comment (iarspider) [inline process-error-reports:1]: Unused script (to automatically open GH issues for failing relvals/tests)?

File: scram-package-monitor-sender [hunk 15]
```diff
@@ -79,7 +79,7 @@ if __name__ == "__main__":
       stops = dict([(x[2], int(x[1])) for x in info if x[0] == "stop"])
       packages = set(x[2] for x in info)
       for x in packages:
-        h = sha1(x + SCRAM_ARCH + CMSSW_VERSION).hexdigest()
+        h = sha1((x + SCRAM_ARCH + CMSSW_VERSION).encode()).hexdigest()
```
Comment (iarspider) [inline scram-package-monitor-sender:82]: Not sure how to test this one...

**Comment by iarspider** [comment]
I can revert scripts that are marked as "probably unused" above.

**Comment by iarspider** [comment]
Documenting script usage:

* [`cache_pull_request.py`](https://github.com/cms-sw/cms-bot/pull/2057/files#diff-5f425f3d05115052a8baa8ede77086bd9c481af7fc4986a78f31721732de3e4f): [cms-prs-cache](https://cmssdt.cern.ch/jenkins/job/cms-prs-cache/), testeed with `python3 cache-pull-request.py  -r cms-sw/cms-bot 2057 2057.json`
* [`cmsdist-comp-pr-process.py`](https://github.com/cms-sw/cms-bot/pull/2057/files#diff-719e257cf81cdec06118769f609ca45bd5cb76331e43c5abe6cfeb5bf12f954a): [comp-bot](https://cmssdt.cern.ch/jenkins/job/comp-bot/), tested with `python3 cmsdist-comp-pr-process.py -n 8670`
* [`cmsdist_merge_permissions.py`](https://github.com/cms-sw/cms-bot/pull/2057/files#diff-0847a638926a430a96865c6bca8781bed480a3f0a26177b82f7d02c121e905e6): library, only used by `cmsdist-comp-pr-process.py`, but the changed function is not used.
* [`das-utils/das_cache.py`](https://github.com/cms-sw/cms-bot/pull/2057/files#diff-6f5382e820c54ac98e081f31ab857828cdb6537f256189fa9aad51dfe3d8dc63):  [das-query](https://cmssdt.cern.ch/jenkins/job/das-query), tested with `git clone git@github.com:cms-sw/cms-sw.github.io.git das_results && python3 das-utils/das_cache.py -c dasgoclient --jobs 16 --threshold 604800 --store das_results/das_queries --query ''`
* [`das-utils/ib-eos-files.py`](https://github.com/cms-sw/cms-bot/pull/2057/files#diff-88f0f057fb4ce565564ca23ced897e29949d0ddd6df7e3cda4482cbb2837d26c): [ib-cache-to-eos](https://cmssdt.cern.ch/jenkins/job/ib-cache-to-eos), tested with `python3 das-utils/ib-eos-files.py --redirector root://cms-xrd-global.cern.ch --file-per-das 0 --jobs 5 --days 4 -n`
* [`es_cmsdoxygen_apache.py`](https://github.com/cms-sw/cms-bot/pull/2057/files#diff-5b6e8e2f6dd5189b02821d6b6e2d00a8ae7d1d0d76d80d82261c7f1b649da323): not used neither in Jenkins, nor in puppet
* [`es_cmsrep_apache.py`](https://github.com/cms-sw/cms-bot/pull/2057/files#diff-1b70276b3f6e59952b52347cfa51efd98319d6d9a649755bd937189ac19032fa): executed by Cron on cmsrep, tested with `python3 es_cmsrep_apache.py` after making `send_payload` a no-op.
*  [`es_cmssdt_apache.py`](https://github.com/cms-sw/cms-bot/pull/2057/files#diff-a93ddab66edfa474d6ecc8c348e4f0a711608da9f341ce93d471c99e4cfe3b98): executed by Cron on cmssdt, tested with `python3 es_cmssdt_apache.py` after making `send_payload` a no-op.
* [`es_doxygen_apache.py`](https://github.com/cms-sw/cms-bot/pull/2057/files#diff-fdf9de506477900f0880c9031a72d072caafe3f8b4a2b617741921dad705c007): not used
* [`es_hypernews_log.py`](https://github.com/cms-sw/cms-bot/pull/2057/files#diff-5e737f946187605b215d9278bc7e3cbcfc79c1694ff7c2f3c5eeb4e7a5022d0a): not used
* [`es_relval_log.py`](https://github.com/cms-sw/cms-bot/pull/2057/files#diff-a16c111fe51d079178fe9fd6f2a9da567d7dbe81821dc34fc5fce83a65161926): not used
* [`forward-pull-requests`](https://github.com/cms-sw/cms-bot/pull/2057/files#diff-001085d67b8fbb5285b4b6fa34485bd43cf06ea4ba118730ff678eb8b47e2291): not used
* [`github_utils.py`](https://github.com/cms-sw/cms-bot/pull/2057/files#diff-ad8c867240c20b825666ccefdbf4d1b4b0d99ae5e93c5b09f9cb9f4d49b937a2): `get_ported_PRs` is unused, tested the other changed functions on my fork of cms-bot
* [`jenkins-jobs/es-cmssw-afs-eos.py`](https://github.com/cms-sw/cms-bot/pull/2057/files#diff-ae8ec4a86ab31f4e7e0a869fc4dccf176296548ee2803dbf536aea91850df663): used by [es-cmssw-afs-eos](https://cmssdt.cern.ch/jenkins/job/es-cmssw-afs-eos), tested by running `python3 jenkins-jobs/es-cmssw-afs-eos.py` (on cmssdt01) after making `send_payload` a no-op
* [`jenkins_monitor_queue.py`](https://github.com/cms-sw/cms-bot/pull/2057/files#diff-aa621a2653c8452b9a33d88f46557323e71c3d1de46c8257a33fc2745671e871): used by [jenkins-monitor-queue](https://cmssdt.cern.ch/jenkins/job/jenkins-monitor-queue), tested (kind of) by running `python3 jenkins_monitor_queue.py` after replacing functions that communicates with ES with no-op. But the information in ES is already corrupted: `Node b"b'\\xe2\\x80\\x98cmsbuild104\\xe2\\x80\\x99"`
* [`logwatch.py`](https://github.com/cms-sw/cms-bot/pull/2057/files#diff-efa928ddcfa9056b0a8c76f8a4774b112107494c2bfd3fc003a8fd5433cf7b5e): Common library for `es-*-apache` scripts
* [`mark_commit_status.py`](https://github.com/cms-sw/cms-bot/pull/2057/files#diff-10c7a25095dc10c1cbce12430d842537cb3f52e98a5969054310892942b00e77): used by [update-pr-status](https://cmssdt.cern.ch/jenkins/job/update-pr-status), tested on my fork of cms-bot
* [`parse_jenkins_builds.py`](https://github.com/cms-sw/cms-bot/pull/2057/files#diff-f7c8884f8bb018dd8f832b2182693c9e2beac82e590231acd037389768a6dbe5): used by [jenkins-elasticsearch-monitor](https://cmssdt.cern.ch/jenkins/job/jenkins-elasticsearch-monitor), tested by running `python3 parse_jenkins_builds.py` after replacing functions that communicates with ES with no-op. 
* [`process-build-release-request`](https://github.com/cms-sw/cms-bot/pull/2057/files#diff-861ba1672783be446e317b4fe880f71c1cafe895b5924a3869b7bc23a1b11fd7): used by [query-build-release-issue](https://cmssdt.cern.ch/jenkins/job/query-build-release-issues), tested with `python3 process-build-release-request -n 42709`
* [`process-error-reports`](https://github.com/cms-sw/cms-bot/pull/2057/files#diff-56adf24cc83a1370893d3d8850bfedeab4065b7f4e54ae86172c3729feab96df): not used
* [`process_pr.py`](https://github.com/cms-sw/cms-bot/pull/2057/files#diff-f37d46a488510bd0b96fd201ff189f8a704b756502df6a1cd46b117986e608a4): library used by many scripts
* [`release-notes`](https://github.com/cms-sw/cms-bot/pull/2057/files#diff-4ae1eb667fb53609046cecfd6c41898b620812e3c1c1128d2f2f74b43bb69a45): used by [release-produce-changelog](https://cmssdt.cern.ch/jenkins/job/release-produce-changelog/), tested with `python3 release-notes -n CMSSW_13_2_1 CMSSW_13_2_2 REL/CMSSW_13_2_1/el8_amd64_gcc11 REL/CMSSW_13_2_2/el8_amd64_gcc11`
* [`release_notes_collection.py`](https://github.com/cms-sw/cms-bot/pull/2057/files#diff-39329062400e50736ea1b5cfe2c9b073166fef464b08629b9a30e204d4183371): used by [update-release-notes-collection](https://cmssdt.cern.ch/jenkins/job/update-release-notes-collection/), tested by `git clone https://github.com/cms-sw/cms-prs && cd ReleaseNotes && python3 ../cms-bot/release_notes_collection.py -r cms-sw/cmssw -p ../cms-prs -N _releases -R CMSSW_13_2_2`
* [`scram-package-monitor-sender`](https://github.com/cms-sw/cms-bot/pull/2057/files#diff-2f32079b6699e357c9959ad4ffa4a7982ec774f778b7b9a2cc8acf210fba50f9): part of SCRAM, I guess, did not test
* [`tag-ib`](https://github.com/cms-sw/cms-bot/pull/2057/files#diff-fac295b4a58eceb90720dc5faaec177de7b25f2e5369efd02a0811a76beb1d73): unused, but tested anyway in dry-run mode.

**Comment by smuzaffar** [comment]
ping

**Comment by smuzaffar** [comment]
test parameters:
- enable = gpu
- skip_tests = static,header

**Comment by smuzaffar** [comment]
please test for el8_amd64_gcc11

**Comment by smuzaffar** [comment]
please test for CMSSW_5_3_X

File: jenkins-jobs/es-cmssw-afs-eos.py [hunk 16]
```diff
@@ -22,7 +22,7 @@
     continue
   week, rel_sec  = cmsswIB2Week (payload["release"])
   payload["@timestamp"]=rel_sec*1000
-  id = sha1("%s-%s-%s" % (payload["release"], payload["architecture"], payload["fstype"])).hexdigest()
+  id = sha1(("%s-%s-%s" % (payload["release"], payload["architecture"], payload["fstype"])).encode()).hexdigest()
```
Comment (smuzaffar) [inline jenkins-jobs/es-cmssw-afs-eos.py:25]: @iarspider , I think the `shebang` statement for this script should be update too ... right?

Comment (iarspider) [reply jenkins-jobs/es-cmssw-afs-eos.py:25]: Fixed

File: es_relval_log.py (hunk 7 above)
Comment (smuzaffar) [reply es_relval_log.py:1]: `runPyRelValThread.py` (which should support both py2 and 3) makes use of it
```
runPyRelValThread.py:from es_relval_log import es_parse_log
```

File: es_utils.py [hunk 17]
```diff
@@ -136,7 +136,7 @@ def get_payload_wscroll(index, query, max_count=-1):
 
 def get_template(index=''):
   data = {'index':index, 'api': '/_template', 'prefix': True}
-  return ssl_urlopen(CMSSDT_ES_QUERY,json.dumps(data))
+  return ssl_urlopen(CMSSDT_ES_QUERY,json.dumps(data).encode())
```
Comment (smuzaffar) [inline es_utils.py:139]: @iarspider , I think we need to `decode` the return value of `ssl_urlopen` e.g. https://github.com/cms-sw/cms-bot/pull/2057/files#diff-403eb1824bda3fe3ae2b63977f99ec9b6b2911b348f16801dc8733e019dea46bL47 
For me the following fails for python3
```
python3 -c 'from es_utils import find_indexes;print(find_indexes("cmssdt-ib-dataset-*"))'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/tmp/cmsbuild/del/cms-bot/es_utils.py", line 143, in find_indexes
    for line in get_indexes(index).split("\n"):
TypeError: a bytes-like object is required, not 'str'
```

File: es_hypernews.py (hunk 5 above)
Comment (smuzaffar) [reply es_hypernews.py:1]: this was used on hypernews server via a cronjob. I think for now we can ignore this as old hypernews server is not in use any more

File: es_doxygen_apache.py (hunk 4 above)
Comment (smuzaffar) [reply es_doxygen_apache.py:1]: right, this was also used by doxygen server which has been deprecated. Now we server cmssw doxygen via CERN IT web infrastructure

File: das-utils/ib-eos-files.py [hunk 18]
```diff
@@ -181,7 +181,10 @@ def copy_lfns_to_eos(eos_lfns):
     while True:
       threads = get_alive_threads(threads)
       if(len(threads) < opts.jobs):
-        log_file=logdir+"/"+sha256(lfn).hexdigest()+".log"
+        if sys.version_info[0] == 3:
+          log_file=logdir+"/"+sha256(lfn.encode()).hexdigest()+".log"
+        else:
+          log_file=logdir+"/"+sha256(lfn).hexdigest()+".log"
```
Comment (smuzaffar) [inline das-utils/ib-eos-files.py:187]: @iarspider , I think we can drop the python2 support from this script

File: das-utils/das_cache.py [hunk 19]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python3
```
Comment (smuzaffar) [inline das-utils/das_cache.py:1]: we can drop python2 support from this

File: cmsdist_merge_permissions.py [hunk 20]
```diff
@@ -44,7 +44,7 @@ def hasRights(user, branch, type, files=[]):
 def isValidWebHook(payload):
```
[the remark this answers is not in the document]
Comment (smuzaffar) [reply cmsdist_merge_permissions.py:44]: this is used via https://github.com/cms-sw/cmssdt-web/blob/master/cgi-bin/github_webhook#L56

**Comment by iarspider** [comment]
There are a few other scripts that use `version_info` to maintain py2 compatibility:
```
buildLogAnalyzer.py
github_scripts/simultaneous_files_modifications_by_PRs.py
github_utils.py
logRootQA.py
report-pull-request-results.py
rucio/install.sh
shift/report.py
splitUnitTestLog.py
```
Should I clean them as well?

**Comment by iarspider** [comment]
> There are a few other scripts that use `version_info` to maintain py2 compatibility:
> 
> ```
> buildLogAnalyzer.py
> github_scripts/simultaneous_files_modifications_by_PRs.py
> github_utils.py
> logRootQA.py
> report-pull-request-results.py
> rucio/install.sh
> shift/report.py
> splitUnitTestLog.py
> ```
> 
> Should I clean them as well?

This can be done in a separate PR, of course.

**Comment by smuzaffar** [comment]
please test for el8_amd64_gcc11

**Comment by smuzaffar** [comment]
please test for CMSSW_5_3_X

**Pull request merged**